### PR TITLE
Use specialization constants in clustered renderer

### LINKED ISSUE #45733: Vulkan: OmniLight3D/SpotLight3D projectors not reimplemented yet
**Godot version:** `master` https://github.com/godotengine/godot/commit/80a4d0fce6a257afd13132918c87f334b4d920c4

**OS/device including version:** Fedora 33, GeForce GTX 1080 (NVIDIA 460.32.03)

**Issue description:**

https://github.com/godotengine/godot/pull/45023 removed light projector support to improve shader occupancy, letting the GPU make better use of parallelism.

reduz said that light projectors should be reimplemented, but placed behind an *opt-in* project setting. The reason for doing this is that projector support increases the forward renderer's base cost, and therefore will slow down rendering even if no projectors are used at all.

**Steps to reproduce:** Look for light projectors in the `master` branch :slightly_smiling_face: 

**Minimal reproduction project:** N/A

### LINKED ISSUE #50094: Vulkan: PCSS-style shadow size options no longer implemented, yet still used in the code
### Godot version

4.0.dev (659178e85)

### System information

Fedora 34, GeForce GTX 1080 (NVIDIA 465.31)

### Issue description

https://github.com/godotengine/godot/pull/45023 removed the PCSS-like features for directional, omni and spot shadows since they were too demanding. However, increasing **Light Size** or **Light Angular Distance** above 0 will still decrease shadow quality noticeably.

We should either reimplement this feature in a more efficient way, or remove the remnants of the old code entirely. I've tried my hand at removing the code [here](https://github.com/Calinou/godot/tree/remove-pcss-soft-shadow-code), but the shader no longer runs because the binding sizes no longer match. I don't know how to fix that.

See also https://github.com/godotengine/godot/issues/50088.

### Steps to reproduce

- Add a DirectionalLight3D node and enable shadows.
- Set the DirectionalLight3D node's **Angular Distance** property to a value above 0, and notice how the shadow quality degrades. Yet, the sky shader will still use this value to enlarge the sun on the procedural sky.
- Add a OmniLight3D node and enable shadows.
- Set the OmniLight3D node's **Shadow Size** property a value above 0, and notice the shadow quality degrading.
- Do the same as OmniLight3D for SpotLight3D.

### Minimal reproduction project

[test_shadow_size.zip](https://github.com/godotengine/godot/files/6756911/test_shadow_size.zip)

### PR DESCRIPTION
**Pull request by reduz**
* Keep track of when projector, soft shadows or directional soft shadows are enabled.
* Enable them via specialization constant where it makes sense.
* Re-implements soft shadows.
* Re-implements light projectors.

image

*Bugsquad edit: Follow-up to https://github.com/godotengine/godot/pull/50325. This closes https://github.com/godotengine/godot/issues/45733 and closes https://github.com/godotengine/godot/issues/50094.*

### CODE REVIEW

**Review by clayjohn — COMMENTED** [review]
Looks good to me!

File: drivers/vulkan/rendering_device_vulkan.cpp [hunk 1]
```diff
@@ -6047,7 +6045,7 @@ RID RenderingDeviceVulkan::render_pipeline_create(RID p_shader, FramebufferForma
 				const PipelineSpecializationConstant &psc = p_specialization_constants[j];
 				if (psc.constant_id == sc.constant.constant_id) {
 					ERR_FAIL_COND_V_MSG(psc.type != sc.constant.type, RID(), "Specialization constant provided for id (" + itos(sc.constant.constant_id) + ") is of the wrong type.");
-					data_ptr[i] = sc.constant.int_value;
+					data_ptr[i] = psc.int_value;
```
Comment (reduz) [inline drivers/vulkan/rendering_device_vulkan.cpp:6048]: took the chance to fix this bug

File: drivers/vulkan/rendering_device_vulkan.cpp [hunk 2]
```diff
@@ -6070,14 +6068,14 @@ RID RenderingDeviceVulkan::render_pipeline_create(RID p_shader, FramebufferForma
...
-				specialization_info.write[k].pMapEntries = specialization_map_entries[k].ptr();
+		for (int i = 0; i < pipeline_stages.size(); i++) {
+			if (specialization_map_entries[i].size()) {
+				specialization_info.write[i].dataSize = specialization_constant_data.size() * sizeof(uint32_t);
+				specialization_info.write[i].pData = data_ptr;
+				specialization_info.write[i].mapEntryCount = specialization_map_entries[i].size();
+				specialization_info.write[i].pMapEntries = specialization_map_entries[i].ptr();
 
-				pipeline_stages.write[k].pSpecializationInfo = specialization_info.ptr();
+				pipeline_stages.write[i].pSpecializationInfo = specialization_info.ptr() + i;
```
Comment (reduz) [inline drivers/vulkan/rendering_device_vulkan.cpp:6078]: this also a bugfix

File: servers/rendering/renderer_rd/forward_clustered/render_forward_clustered.cpp [hunk 3]
```diff
@@ -318,11 +318,11 @@ void RenderForwardClustered::_render_list_template(RenderingDevice::DrawListID p
 	RID prev_pipeline_rd;
 	RID prev_xforms_uniform_set;
 
-	bool shadow_pass = (p_params->pass_mode == PASS_MODE_SHADOW) || (p_params->pass_mode == PASS_MODE_SHADOW_DP);
+	bool shadow_pass = (p_pass_mode == PASS_MODE_SHADOW) || (p_pass_mode == PASS_MODE_SHADOW_DP);
```
Comment (reduz) [inline servers/rendering/renderer_rd/forward_clustered/render_forward_clustered.cpp:321]: these should be using p_pass_mode which is resolved at compile time for more performance., so replaced it.

**Comment by BastiaanOlij** [comment]
Need to look at this more closely but sounds like a good way to improve multiview as well, instead of using defines add a constant in. Though not sure if that would rhyme well with us not wanting to compile any multiview shaders if XR is not enabled.

**Comment by reduz** [comment]
@BastiaanOlij specialization constants work as long as you dont add any new input or output to the shaders, not sure if this is the case with multiview, but probably not.

**Review by clayjohn — APPROVED** [review]

**Comment by BastiaanOlij** [comment]
> 
> 
> @BastiaanOlij specialization constants work as long as you dont add any new input or output to the shaders, not sure if this is the case with multiview, but probably not.

Hmmm, thats a good point because Multiview adds the gl_ViewIndex built-in input.

**Pull request merged**